### ISSUE TITLE
Scroll wheel zoom works on Win10 but not Win7

### ISSUE THREAD

**Issue by LogDogg** [swharden/ScottPlot#354]
Bug description:
Scroll wheel zoom not working in WinForms, enableScrollWheelZoom is set to true.

Steps to reproduce:
1. Drag and drop a formsPlot into the designer.
2. Enable the scroll with zoom using
`formsPlot1.Configure(enableScrollWheelZoom: true);`
3. Start the program and try to zoom using the scroll wheel

Additional info:
Everything else works as expected. You can zoom using right click. Using .Configure works on other settings properly. MouseWheel event is working properly. Everything works fine in the WPF version, this is a problem with the WinForms version. I'm using the 2017 version of Visual Studio with the latest version of ScottPlot on NuGet (4.0.29).

**Comment by LogDogg** [comment]
I've solved it. It was an issue with PictureBox focus. I just added
`pbPlot.MouseEnter += (s, e) => pbPlot.Focus();`
under `InitializeScottPlot()` in the FormsPlot.cs User Control. It now works as intended.

**Comment by LogDogg** [comment]
Hey @swharden, I'm gonna leave this issue open just so you can see it and apply the changes if necessary.

**Comment by swharden** [comment]
Hi @LogDogg thanks for reporting this!

I'm trying to recreate your issue but I can't seem to find any situation where the scroll wheel doesn't zoom. I created a textbox and even as I type in the textbox (so it's clearly in focus) the plot responds to the scrollwheel.

![image](https://user-images.githubusercontent.com/4165489/80928414-3993d800-8d72-11ea-95c6-aba3ed84106b.png)

Even right now with my Chrome window selected (the windows form window isn't even selected) the scroll wheel zoom still works.

Can you give me more information to help me recreate this issue?

Thanks!

**Comment by swharden** [comment]
(I'll close this issue until I learn more about it or gain the ability to recreate it)

**Comment by HemetsS** [comment]
Hi @swharden,

i just had the same issue as @LogDogg.
Maybe it´s a kind of a windows thing?
I am using Windows 7 Pro.
![grafik](https://user-images.githubusercontent.com/48977894/99255523-55ddfa80-2814-11eb-949a-c8093afcb3cc.png)
When I am using the same program on a Windows 10 machine, it works.

I used the current Master Version 4.0.42.
The problem was also visible in the examples (ScottPlot-master\src\ScottPlot.Demo.WinForms).

After I was adding the workaround as @LogDogg  mentioned in a previous comment it was working.

**Comment by vrdriver** [comment]
Hi Scott, I can also confirm that scrolling is not working in Windows 7 Pro SP1 x64 for me either, both in a VM and a physical machine (over TeamViewer). I have been developing in Windows 10 Pro and haven't had any dramas. It's only now that I'm trying other versions that I've just found this.

The following packages are installed:

```
<?xml version="1.0" encoding="utf-8"?>
<packages>
  <package id="NAudio" version="1.10.0" targetFramework="net452" />
  <package id="Newtonsoft.Json" version="12.0.3" targetFramework="net452" />
  <package id="ScottPlot" version="4.0.45" targetFramework="net461" />
  <package id="ScottPlot.WinForms" version="4.0.45" targetFramework="net461" requireReinstallation="true" />
  <package id="System.Drawing.Common" version="5.0.0" targetFramework="net461" />
  <package id="System.ValueTuple" version="4.5.0" targetFramework="net472" />
  <package id="z440.atl.core" version="3.14.0" targetFramework="net461" />
</packages>
```

This is presently the initialisation code:
    
```
formsPlot2.plt.Style(ScottPlot.Style.Gray1);
formsPlot2.plt.AxisBounds(0, 10, -1, 1);
formsPlot2.plt.AxisAuto(0); 
formsPlot2.Configure(lockVerticalAxis: true, enableRightClickMenu: false, enableDoubleClickBenchmark: false);
formsPlot2.plt.PlotSignal(dataX1,   color: Color.SkyBlue); // Makes it draw
formsPlot2.plt.PlotSignal(dataX2,   color: Color.LightBlue); // Makes it draw
formsPlot2.plt.Ticks(rulerModeX: true, displayTicksY: false, color: Color.White, dateTimeX: true, dateTimeFormatStringX: "HH:mm:ss.fff"); 
formsPlot2.plt.AxisAuto(tightenLayout: true);
formsPlot2.plt.Frame(drawFrame: false, left: false, right: false, top: false, bottom: false);
formsPlot2.plt.TightenLayout(padding: 0, render: true);
formsPlot2.Configure(enableScrollWheelZoom: true, recalculateLayoutOnMouseUp: false); 
formsPlot2.Render(); 
```


Also, I have these added too:
```
this.formsPlot2.MouseMoved += new System.Windows.Forms.MouseEventHandler(this.formsPlot2_MouseMoved_1);
this.formsPlot2.MouseUp += new System.Windows.Forms.MouseEventHandler(this.formsPlot2_MouseUp);
this.formsPlot2.MouseWheel += new System.Windows.Forms.MouseEventHandler(this.formsPlot2_MouseWheel);
this.formsPlot2.MouseDown += new System.Windows.Forms.MouseEventHandler(this.formsPlot2_MouseDown);
this.formsPlot2.MouseDoubleClick += new System.Windows.Forms.MouseEventHandler(this.formsPlot2_MouseDown);
this.formsPlot2.MouseDragPlottable += (s, e) => { formsPlot2_MouseDragPlottable(s, e); };

```

I even tried the above mentioned code
`this.formsPlot2.MouseEnter += (s, e) => this.formsPlot2.Focus();`

But, at the moment, I haven't seen a solution just yet.

**Comment by swharden** [comment]
Windows is full of surprises! I think this issue is described [here on stackoverflow](https://stackoverflow.com/questions/44735362/control-mousewheel-event): _Win 10 has a system option called "Scroll inactive windows when I hover over them”. That's why my previous code is working only in Win 10 machine._

@vrdriver this makes me think if you do something like this, it may work:

```cs
this.formsPlot2.MouseHover+= (s, e) => this.formsPlot2.Focus();
```

What you tried was pretty close! Want to give this one a try and see if it fixes your problem?

**Comment by vrdriver** [comment]
Hi Scott, 
I've now fired up VS2019 a Win7 VM environment so I can test my code a little better.

I've now added this as per the suggestion:
`this.formsPlot2.MouseHover+= (s, e) => this.formsPlot2.Focus();`

I am also capturing the MouseWheel so I can do other processes
`this.formsPlot2.MouseWheel += new System.Windows.Forms.MouseEventHandler(this.formsPlot2_MouseWheel);`

For the sake of this example, I also did some checks.

```
public void formsPlot2_MouseWheel(object sender, MouseEventArgs e)
{
          if (!formsPlot2.Focused)
          {
                formsPlot2.Focus();
                Console.WriteLine("Now focused");
           }
           else
           {
                Console.WriteLine("Was already focused");
            }
           Console.WriteLine("e Delta:" + e.Delta);
}
```

However, what I'm getting back in the log is:

```
Was already focused
e Delta:88
Was already focused
e Delta:88
Was already focused
e Delta:120
Was already focused
e Delta:-120
Was already focused
e Delta:-154
Was already focused
e Delta:-154
Was already focused
e Delta:-331
Was already focused
e Delta:-331
Was already focused
e Delta:-177
Was already focused
e Delta:-177
Was already focused
e Delta:-120

etc...
```

I'm just guessing here, but it seems to be capturing the scroll wheel correctly, but SP may not be actioning the scroll wheel properly?

**Comment by swharden** [comment]
This might be a bit of a hack, but if you are already intercepting scroll wheel events maybe you could manually adjust the axis to respond to them?

```cs
public void formsPlot2_MouseWheel(object sender, MouseEventArgs e)
{
	double xFrac = (e.Delta > 0) ? 1.15 : 0.85;
	double yFrac = (e.Delta > 0) ? 1.15 : 0.85;
	double zoomToX = formsPlot1.plt.CoordinateFromPixelX(e.Location.X);
	double zoomToY = formsPlot1.plt.CoordinateFromPixelY(e.Location.Y);
	formsPlot1.plt.AxisZoom(xFrac, yFrac, zoomToX, zoomToY);
	formsPlot1.Render();
}
```

**Comment by vrdriver** [comment]
Hi Scott, Thanks for pointing me once again in the right direction! I was worried that doing that manual interception would give Windows10 a double scroll, but fortunately it didn't do that.

I've made some great progress!

Even though what you are calling a hack, let's just call it a work-around instead. I experimented more and took it a little further by checking the OS version as well, because I found that scrolling on Windows10 worked faster than Windows 7.

So, to separate the 2 scenarios , I've implemented the following (as a first play):

```
// https://stackoverflow.com/questions/2819934/detect-windows-version-in-net
        string getOSInfo()
        {
            //Get Operating system information.
            OperatingSystem os = Environment.OSVersion;
            //Get version information about the os.
            System.Version vs = os.Version;

            //Variable to hold our return value
            string operatingSystem = "";

            if (os.Platform == PlatformID.Win32Windows)
            {
                //This is a pre-NT version of Windows
                switch (vs.Minor)
                {
                    case 0:
                        {
                            //operatingSystem = "95";
                            operatingSystem = "0";
                        }
                        break;
                    case 10:
                        if (vs.Revision.ToString() == "2222A")
                        {
                            //operatingSystem = "98SE";
                            operatingSystem = "1";
                        }
                        else
                        {
                            //operatingSystem = "98";
                            operatingSystem = "2";
                        }
                        break;
                    case 90:
                        {
                            //operatingSystem = "Me";
                            operatingSystem = "3";
                        }
                        break;
                    default:
                        break;
                }
            }
            else if (os.Platform == PlatformID.Win32NT)
            {
                switch (vs.Major)
                {
                    case 3:
                        //operatingSystem = "NT 3.51";
                        operatingSystem = "3.51";
                        break;
                    case 4:
                        //operatingSystem = "NT 4.0";
                        operatingSystem = "4.0";
                        break;
                    case 5:
                        if (vs.Minor == 0)
                        {
                            //operatingSystem = "2000";
                            operatingSystem = "5.0";
                        }
                        else
                        {
                            //operatingSystem = "XP";
                            operatingSystem = "5.1";
                        }
                        break;
                    case 6:
                        if (vs.Minor == 0)
                        {
                            //operatingSystem = "Vista";
                            operatingSystem = "6";
                        }
                        else if (vs.Minor == 1)
                            operatingSystem = "7";
                        else if (vs.Minor == 2)
                            operatingSystem = "8";
                        else
                            operatingSystem = "8.1";
                        break;
                    case 10:
                        operatingSystem = "10";
                        break;
                    default:
                        break;
                }
            } 
            //Return the information we've gathered.
            return operatingSystem;
        }




        public void formsPlot2_MouseWheel(object sender, MouseEventArgs e)
        {
            if (!formsPlot2.Focused)
            {
                formsPlot2.Focus();
                Console.WriteLine("Now focused");
            }
            else
            {
                Console.WriteLine("Was already focused");
            }
             
            Console.WriteLine("e Delta:" + e.Delta);
 

            // This is a fix for Windows 7 and earlier.
 
            // It's seems slightly slower, but it brings back scrollable zooming.
            if (Convert.ToInt16(getOSInfo()) < 8)
            {
                double xFrac = (e.Delta > 0) ? 1.15 : 0.85;
                //double yFrac = (e.Delta > 0) ? 1.15 : 0.85;
                double zoomToX = formsPlot2.plt.CoordinateFromPixelX(e.Location.X);
                //double zoomToY = formsPlot2.plt.CoordinateFromPixelY(e.Location.Y);
                formsPlot2.plt.AxisZoom(xFrac, 1, zoomToX);
                formsPlot2.Render(); 
                formsPlot2.Configure(lockVerticalAxis: true);  
                formsPlot2.Focus(); 
            }
}

```

I also discovered if you don't call the Render() method after the delta scroll that you have to click on the Plot each time before it will respond as well. I hope this butchered code helps someone else. 

Off to clean it up now...

**Comment by swharden** [comment]
Thanks @vrdriver! I cleaned it up a bit and will leave it here for future people who may search for this similar question. I'm happy to hear you found a solution that works for you!

```cs
public void formsPlot2_MouseWheel(object sender, MouseEventArgs e)
{
    // Windows 7 and earlier need this to support scroll wheel zooming
    if (Environment.OSVersion.Version.Major <= 7)
    {
        formsPlot2.Focus();
        double xFrac = (e.Delta > 0) ? 1.15 : 0.85;
        double yFrac = (e.Delta > 0) ? 1.15 : 0.85;
        double zoomToX = formsPlot2.plt.CoordinateFromPixelX(e.Location.X);
        double zoomToY = formsPlot2.plt.CoordinateFromPixelY(e.Location.Y);
        formsPlot2.plt.AxisZoom(xFrac, yFrac, zoomToX, zoomToY);
        formsPlot2.Render();
        formsPlot2.Focus();
    }
}
```

**Comment by swharden** [comment]
I couldn't leave that stack overflow code block alone though 😅 

Here's a more straightforward version

```cs
/// <summary>
/// Return the version and name of the Windows operating system in use
/// </summary>
public static (decimal version, string name) GetOsInfo()
{
    // Derived from https://stackoverflow.com/questions/2819934/detect-windows-version-in-net
    int minorVersion = Environment.OSVersion.Version.Minor;
    int majorVersion = Environment.OSVersion.Version.Major;

    if (minorVersion == 0)
        return (0, "Windows 95");
    if (minorVersion == 10)
        return (2, "Windows 98");
    if (minorVersion == 90)
        return (3, "Windows ME");
    if (majorVersion == 3)
        return ((decimal)3.51, "Windows NT3");
    if (majorVersion == 4)
        return ((decimal)4.0, "Windows NT4");
    if (majorVersion == 5 && minorVersion == 0)
        return ((decimal)5.0, "Windows 2000");
    if (majorVersion == 5)
        return ((decimal)5.1, "Windows XP");
    if (majorVersion == 6 && minorVersion == 0)
        return (6, "Windows Vista");
    if (majorVersion == 6 && minorVersion == 1)
        return (7, "Windows 7");
    if (majorVersion == 6 && minorVersion == 2)
        return (8, "Windows 8");
    if (majorVersion == 6)
        return ((decimal)8.1, "Windows 8.1");
    if (majorVersion == 10)
        return (10, "Windows 10");

    throw new InvalidOperationException("Unknown Windows Version");
}
```

**Comment by vrdriver** [comment]
Yes, I know the problem with SO. You see an idea that works, and you mould and shape. That's a pretty great summary. I'll probably take it a step further and just make it a global variable and so not have to calculate this every time there is a mousewheel scroll. Keep up the great work. Thanks again for the pointers.

**Comment by swharden** [comment]
Sure thing!

>  I'll probably take it a step further and just make it a global variable

Actually, that may not be required. This if statement is pretty clean, and doesn't require any other functions:

```cs
public void formsPlot2_MouseWheel(object sender, MouseEventArgs e)
{
    // Windows 7 and earlier need this to support scroll wheel zooming
    if (Environment.OSVersion.Version.Major <= 7)
    {
        formsPlot2.Focus();
        double xFrac = (e.Delta > 0) ? 1.15 : 0.85;
        double yFrac = (e.Delta > 0) ? 1.15 : 0.85;
        double zoomToX = formsPlot2.plt.CoordinateFromPixelX(e.Location.X);
        double zoomToY = formsPlot2.plt.CoordinateFromPixelY(e.Location.Y);
        formsPlot2.plt.AxisZoom(xFrac, yFrac, zoomToX, zoomToY);
        formsPlot2.Render();
        formsPlot2.Focus();
    }
}
```

**Comment by vrdriver** [comment]
Yes. That statement is **very** clean. Love it. 

But there's a small note on [here](https://docs.microsoft.com/en-au/windows/win32/api/winnt/ns-winnt-osversioninfoexa?redirectedfrom=MSDN) that states the following:

_* For applications that have been manifested for Windows 8.1 or Windows 10. Applications not manifested for Windows 8.1 or Windows 10 will return the Windows 8 OS version value (6.2). To manifest your applications for Windows 8.1 or Windows 10, refer to Targeting your application for Windows._ 

It's also written out here (to be more understandable) as well with the important note. [https://stackoverflow.com/a/2819962/1190051](https://stackoverflow.com/a/2819962/1190051)

_Important note: if your executable assembly manifest doesn't explicitly state that your exe assembly is compatible with Windows 8.1 and Windows 10.0, System.Environment.OSVersion will return Windows 8 version, which is 6.2, instead of 6.3 and 10.0!_
  
In short though, the revision you did on the OS version checker routine will probably more than cater for this in the future. 

In reality, I was just more concerned on the hit of system checking each time it scrolled.

**Comment by swharden** [comment]
Thanks for the details! There's definitely some nuance there. Windows versioning is unexpectedly complicated 😩 

```cs
bool isWin7orNewer = (Environment.OSVersion.Version.Major >= 7) ||
                     ((Environment.OSVersion.Version.Major == 6) && (Environment.OSVersion.Version.Minor > 0));
```
